### ISSUE TITLE
Implement safe navigation operator

### ISSUE THREAD

**Issue by Simn** [HaxeFoundation/haxe#10479]
See https://github.com/HaxeFoundation/haxe-evolution/pull/89

**Comment by Simn** [comment]
After starting to implement this, I wonder if we should do something for cases like `expr?.field(arg)`, or even `expr?.field.field`. I know these things would be covered by null-safety, but in my mind constructs like these are very questionable. It makes no sense to me to follow up a safe access with any kind of unsafe one, because that makes the entire thing unsafe by definition.

Am I overthinking this?

**Comment by back2dos** [comment]
> It makes no sense to me to follow up a safe access with any kind of unsafe one, because that makes the entire thing unsafe by definition.

Depends. If `expr?.field` is an abstract that handles `this == null` internally, then further operations would be safe. In the same vein, static extensions that deal with the null case would work just fine.

This makes for totally normal code:

```haxe
function isEmpty(s:Null<String>) return s == null || s.trim() == '';

var user:Null<{ name:String }> = null;
if (user?.name.isEmpty())
  trace('no name supplied');
```

You can either say that it's the users problem to use this feature sensibly. Or you can prevent all potentially nonsensical use and then exempt valid use cases as discovered. The latter will lead to more robust behavior, but it'll require more effort and I'm not sure that's worth it when we already have null safety.

**Comment by Simn** [comment]
Static extensions are a good point, I suppose.

Regarding abstracts, I don't think this feature should trigger any `==` overload. The fact that this transforms to a `== null` check is an implementation detail. We might even want to pipe `?.` through to target generators that support it later for cleaner output code.

**Comment by back2dos** [comment]
I wasn't referring to overloading. What I meant is this:

```haxe
abstract Wrapper(Null<Actual>) {
  public function foo()
    if (this == null) fallbackBehavior();
    else this.bar();
}
```

**Comment by Simn** [comment]
This has been implemented in #10561. Further work will be tracked in other issues.
